tensorflow_quantum/python/optimizers/rotosolve_minimizer.py

```diff
@@ -0,0 +1,284 @@
+    """Return static shape of tensor `x` if available,
+     else `tf.shape(x)`.
+    Args:
+      x: `Tensor` (already converted).
+    Returns:
+      Numpy array (if static shape is obtainable), else `Tensor`.
+    """
+    return prefer_static_value(tf.shape(x))
+
+
+def prefer_static_value(x):
+    """Return static value of tensor `x` if available, else `x`.
+    Args:
+      x: `Tensor` (already converted).
+    Returns:
+      Numpy array (if static value is obtainable), else `Tensor`.
+    """
+    static_x = tf.get_static_value(x)
+    if static_x is not None:
+        return static_x
+    return x
+
+
+RotosolveOptimizerResults = collections.namedtuple(
+    'RotosolveOptimizerResults',
+    [
+        'converged',  # Scalar boolean tensor indicating whether the minimum
+        # was found within tolerance.
+        'num_iterations',  # The number of iterations of the BFGS update.
+        # The total number of objective evaluations performed.
+        'num_objective_evaluations',
+        'position',  # A tensor containing the last argument value found
+        # during the search. If the search converged, then
+        # this value is the argmin of the objective function.
+        # A tensor containing the value of the objective from previous iteration
+        'last_objective_value',
+        'objective_value',  # A tensor containing the value of the objective
+        # function at the `position`. If the search
+        # converged, then this is the (local) minimum of
+        # the objective function.
+        'tolerance',  # Define the stop criteria. Iteration will stop when the
+        # objective value difference between two iterations is smaller than
+        # tolerance
+        'solve_param_i',  # The parameter index where rotosolve is currently
+        # modifying. Reserved for internal use.
+    ])
+
+
+def _get_initial_state(initial_position, tolerance, expectation_value_function):
+    """Create RotosolveOptimizerResults with initial state of search ."""
+    init_args = {
+        "converged": tf.Variable(False),
+        "num_iterations": tf.Variable(0),
+        "num_objective_evaluations": expectation_value_function.iter,
+        "position": tf.Variable(initial_position),
+        "objective_value": tf.Variable(0.),
+        "last_objective_value": tf.Variable(0.),
+        "tolerance": tolerance,
+        "solve_param_i": tf.Variable(0)
+    }
+    return RotosolveOptimizerResults(**init_args)
+
+
+def minimize(expectation_value_function,
+             initial_position,
+             tolerance=1e-8,
+             max_iterations=50,
+             name=None):
+    """Applies the rotosolve algorithm to minimize a linear combination
+     of quantum measurement expectation values. See arXiv:1903.12166,
+      arXiv:1905.09692
+
+    ### Usage:
+
+    The following example demonstrates the Rotosolve optimizer attempting
+     to find the minimum for two qubit ansatz expectation value.
+
+    ```python

```

Comment:
This is a really good useage example. It is a little too complex for something that we want on the API docs. Do you think we could shorten things up a little bit ?

Also can you use `>>>` and `...` like in https://github.com/tensorflow/quantum/blob/master/tensorflow_quantum/python/layers/high_level/pqc.py ?

Reply:
Updated in 41b3161fc10974521e3223b2bc63a4cd0d6d1a23. Any changes suggested?